rfcs/006-package-management.md

```diff
@@ -0,0 +1,560 @@
+version of a specific package). nixpkgs works similarly; a single revision
+of nixpkgs fixes the versions of all the packages in nixpkgs.
+
+We might have some trouble automatically coming up with mutually compatible
+snapshots, though. Haskell gets to use compile-time checks to test compatibility,
+and nixpkgs just does a lot of building and testing to check whether everything
+works. Nickel being dynamic and lazy might make this hard.
+
+### Alternative: minimal version selection
+
+Go uses a system called "[minimal version selection]" in which all version
+updates are required to be backwards-compatible: if you want to break some
+API, you have to add the new API at a different path and retain the old API.
+Then version resolution is trivial: just take the oldest version that is
+compatible with all the requirements. Advantages:
+
+- version resolution is trivial
+- version specification is trivial (there are no version ranges, just lower bounds)
+- version resolution is stable over time, even without a lock file; resolving
+  to the oldest compatible version means that we don't change resolution just
+  because a new version became available
+
+I don't have much experience using this method, but it seems like one
+disadvantage is that it puts an extra burden on package maintainers to keep
+backwards compatibility even across "major" version bumps.
+
+[bzlmod] also uses some version of minimal version selection, but they
+do apparently allow for backwards-compatibility breakage. They have
+an extra piece of metadata called "compatibility level" (not encoded in
+the version number). They don't allow a dependency graph to contain different
+compatibility levels of the same module.
+
+[minimal version selection]: https://research.swtch.com/vgo-mvs
+[bzlmod]: https://docs.bazel.build/versions/5.0.0/bzlmod.html
+
+### Some interpolation between MVS and semver
+
+There are some middle grounds between "minimal version selection" and
+cargo-style resolution. We could follow semver and allow multiple incompatible
+versions of packages, while also doing one or both of the following
+
+- preferring to resolve minimal versions instead of maximal versions. This promotes
+  stability in version resolution, at the cost of requiring intervention to bring
+  in bugfixes in dependencies.
+- only allowing `^` version specifications (along with `=` overrides at the top level).
+  This amounts to being opinionated that "the ecosystem is using semver, and your
+  version requirements should comply with semver." It's basically a semver analogue
+  of go's policy of only allowing lower bounds (plus pinning at the top level).
+
+### Question: how to handle pre-1.0 minor versions?
+
+Officially, semver says that pre-1.0 versions are mutually incompatible. If we
+follow this, pre-1.0 versions would never get binned together. `cargo` modifies
+the semver rules, allowing `0.x.y` versions to be binned together if they share
+the same `x`. Should we do the same?
```

Comment:
I think so. Otherwise I don't understand why pre-1.0 version have both a minor and a patch version, if they don't have any useful distinction semantically.